docs/customizations.md

```diff
@@ -0,0 +1,233 @@
+    def operation(self, body: Dict[str, Any], /, **kwargs: Any):
+        """Pass in the body as a positional only parameter."""
+
+    @overload
+    def operation(self, *, foo: str, bar: str, **kwargs: Any):
+        """Pass in the body as splatted keyword only arguments."""
+
+    def operation(self, *args, **kwargs):
+        """Base operation for the two overloads"""
+        if not args:
+            args.append({"foo": kwargs.pop("foo"), "bar": kwargs.pop("bar")})
+        return super().operation(*args, **kwargs)
+
+__all__ = ["OperationGroup"]
+```
+
+### Change Client Behavior
+
+In this example, we change the default authentication policy for a client.
+
+In this `_patch.py` file:
+
+```
+azure-sdk
+│   README.md
+│
+└───azure
+    └───sdk
+        │   _service_client.py # where the generated service client is
+        |   _patch.py # where we customize the models code
+        └───operations
+        └───models
+```
+
+```python
+from ._service_client import ServiceClient as ServiceClientGenerated
+from azure.core.pipeline import PipelineRequest
+from azure.core.pipeline.policies import SansIOHTTPPolicy
+
+class MyAuthenticationPolicy(SansIOHTTPPolicy):
+
+    def __init__(self, key: str):
+        self.key = key
+
+    def on_request(self, request: PipelineRequest):
+        request.http_request.headers["Authorization"] = f"My key is {self.key}"
+        return super().on_request(request)
+
+class ServiceClient(ServiceClientGenerated):
+
+    def __init__(self, endpoint: str, credential: str, **kwargs):
+        super().__init__(
+            endpoint=endpoint,
+            credential=credential,
+            authentication_policy=kwargs.pop("authentication_policy", MyAuthenticationPolicy(credential)),
+            **kwargs
+        )
+
+__all__ = ["ServiceClient"]
+```
+
+### Add a Client Method
+
+Similar to models and operations, you can override client behavior in a `_patch.py` file, this time
+at the root of the sdk.
+
+Here, we will be adding an alternate form of authentication on the client, class method `from_connection_string`.
+
+In this `_patch.py` file:
+
+```
+azure-sdk
+│   README.md
+│
+└───azure
+    └───sdk
+        │   _service_client.py # where the generated service client is
+        |   _patch.py # where we customize the models code
+        └───operations
+        └───models
+```
+
+```python
+from typing import Any
+from azure.core.credentials import AzureKeyCredential
+from ._service_client import ServiceClient as ServiceClientGenerated
+
+class ServiceClient(ServiceClientGenerated):
+
+    @classmethod
+    def from_connection_string(cls, connection_string: str, **kwargs: Any):
+        parsed_connection_string = _parse_connection_string(connection_string) # parsing function you've defined
+        return cls(
+            credential=AzureKeyCredential(parsed_connection_string.pop("accesskey")),
+            endpoint=parsed_connection_string.pop("endpoint")
+        )
+
+__all__ = ["ServiceClient"]
+```
```

Comment:
We could add more doc reference at the end of doc for users ^^: https://github.com/Azure/azure-sdk-for-python/blob/main/doc/dev/customize_code/how-to-patch-sdk-code.md

Reply:
Sounds good, I can include your `patch_sdk()` tips at the bottom! One thing to note though: with the new customizations design, we want to leave `patch_sdk()` as a last resort for SDK developers, i.e. if customizations doesn't work for them, then they use `patch_sdk()`. That's because `patch_sdk()` does not have good typehinting for SDK writers and for end users.

Also, I just learned this myself, but there are a bunch of pre-defined directive functions [here](https://github.com/Azure/autorest/blob/main/docs/generate/built-in-directives.md) that make renaming much easier, so I don't think we need to include any rename examples in the `how-to-patch-sdk-code` doc anymore!